### PR TITLE
Cleanup dynamic fee windower logic

### PR DESCRIPTION
**Pull request by StephenButtolph**
## Why this should be merged

The existing windowing logic is overly complex.

1. The format of the window is baked into `CalcBaseFee`.
2. The implementation supports arbitrary sized window elements even though we never use this. Specifically how we have `rollWindow` and `rollLongWindow` even though all we ever use are `uint64`s.

## How this works

Moves all the windower logic into a new type `DynamicFeeWindow`

## How this was tested

Expanded unit tests (and existing dynamic fees tests)

## Need to be documented?

no.

## Need to update RELEASES.md?

no.

### CODE REVIEW

File: consensus/dummy/dynamic_fees.go [hunk 1]
```diff
@@ -83,65 +69,56 @@ func CalcBaseFee(config *params.ChainConfig, parent *types.Header, timestamp uin
...
 
-		// Only add the [blockGasCost] to the gas used if it isn't AP5
-		if !isApricotPhase5 {
-			addedGas, overflow = math.SafeAdd(addedGas, blockGasCost)
-			if overflow {
-				addedGas = math.MaxUint64
-			}
-		}
+	// Compute the new state of the gas rolling window.
+	addedGas := add(parent.GasUsed, parentExtraStateGasUsed, blockGasCost)
```
Comment (StephenButtolph) [inline consensus/dummy/dynamic_fees.go:105]: This isn't a change in behavior, but it feels odd to me that we added the `blockGasCost` to the parent's `addedGas` rather than the current blocks.

File: consensus/dummy/dynamic_fees.go [hunk 2]
```diff
@@ -83,65 +69,56 @@ func CalcBaseFee(config *params.ChainConfig, parent *types.Header, timestamp uin
...
+	dynamicFeeWindow.Add(addedGas)
 
-		slot := params.RollupWindow - 1 - roll
-		start := slot * wrappers.LongLen
-		updateLongWindow(newRollupWindow, start, addedGas)
```
Comment (StephenButtolph) [inline consensus/dummy/dynamic_fees.go:137]: This slot management and surrounding `if` was only needed because we decided to shift at the start of the function rather than later in the function. By moving the shift, we can avoid all these checks.

File: consensus/dummy/dynamic_fees.go [hunk 3]
```diff
@@ -214,76 +191,6 @@ func selectBigWithinBounds(lowerBound, value, upperBound *big.Int) *big.Int {
...
-// [1, 2, 3, 4]
-// Roll = 1
-// [2, 3, 4, 0]
-// Roll = 2
-// [3, 4, 0, 0]
-// Roll = 3
-// [4, 0, 0, 0]
-// Roll >= 4
-// [0, 0, 0, 0]
-// Assumes that [roll] is greater than or equal to 0
```
Comment (StephenButtolph) [inline consensus/dummy/dynamic_fees.go:232]: The new code uses a `uint64` to avoid any assumptions (which actually flows more naturally anyways, because we actually cast from a `uint64 -> int` and then assume the `int >= 0` rather than just using the `uint64`.

File: consensus/dummy/dynamic_fees.go [hunk 4]
```diff
@@ -214,76 +191,6 @@ func selectBigWithinBounds(lowerBound, value, upperBound *big.Int) *big.Int {
...
-// If an overflow occurs, while summing the contents, the maximum uint64 value is returned.
-func sumLongWindow(window []byte, numLongs int) uint64 {
-	var (
-		sum      uint64 = 0
-		overflow bool
-	)
-	for i := 0; i < numLongs; i++ {
-		// If an overflow occurs while summing the elements of the window, return the maximum
-		// uint64 value immediately.
-		sum, overflow = math.SafeAdd(sum, binary.BigEndian.Uint64(window[wrappers.LongLen*i:]))
```
Comment (StephenButtolph) [inline consensus/dummy/dynamic_fees.go:266]: It seems like there was a decent amount of work to avoid parsing the window into a `[]uint64`, but that's actually what we are doing here anyways... So assuming overflow is infrequent (which it should never happen in reality) just using a `[...]uint64` should be just as efficient but avoid the `[]byte` handling.

File: consensus/dummy/dynamic_fees.go [hunk 5]
```diff
@@ -83,65 +69,56 @@ func CalcBaseFee(config *params.ChainConfig, parent *types.Header, timestamp uin
...
+		// may not have a populated [ExtDataGasUsed].
+		if parent.ExtDataGasUsed != nil {
+			parentExtraStateGasUsed = parent.ExtDataGasUsed.Uint64()
 		}
+	default:
+		blockGasCost = ApricotPhase3BlockGasFee
+	}
 
-		// Only add the [blockGasCost] to the gas used if it isn't AP5
-		if !isApricotPhase5 {
```
Comment (StephenButtolph) [inline consensus/dummy/dynamic_fees.go:128]: This check isn't needed, as `blockGasCost` is `0` if `isApricotPhase5` is true.

File: consensus/dummy/dynamic_fee_window.go [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+		window[i] = binary.BigEndian.Uint64(bytes[offset:])
+	}
+	return window, nil
+}
+
+// Add adds the amount to the most recent entry in the window.
+//
+// If the most recent entry overflows, it is set to [math.MaxUint64].
+func (w *DynamicFeeWindow) Add(amount uint64) {
+	const lastIndex uint = params.RollupWindow - 1
```
Comment (StephenButtolph) [inline consensus/dummy/dynamic_fee_window.go:46]: This was something fun I discovered working on the canoto stuff, if `params.RollupWindow` is modified to make this code panic, by defining it as a `const ... uint`, this code actually won't compile.

Not really important for this code tbh... But I like it 🤷

File: consensus/dummy/dynamic_fee_window_test.go [hunk 7]
```diff
@@ -0,0 +1,246 @@
...
+
+func TestDynamicFeeWindow_Add(t *testing.T) {
+	tests := []struct {
+		name     string
+		window   DynamicFeeWindow
+		amount   uint64
+		expected DynamicFeeWindow
+	}{
+		{
+			name: "normal addition",
```
Comment (darioush) [inline consensus/dummy/dynamic_fee_window_test.go:23]: nit: seems preferable to use `_` in the names passed to t.Run, since spaces are replaced with `_`, then we can search any failing test by string in the code.

**Review by darioush — APPROVED** [review]
nice cleanup

File: consensus/dummy/dynamic_fee_window.go [hunk 8]
```diff
@@ -0,0 +1,92 @@
+// (c) 2019-2025, Ava Labs, Inc. All rights reserved.
```
Comment (qdm12) [inline consensus/dummy/dynamic_fee_window.go:1]: nit shouldn't this be:

```suggestion
// (c) 2025, Ava Labs, Inc. All rights reserved.
```

since it's a new file

File: consensus/dummy/dynamic_fee_window.go [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+	"encoding/binary"
+	"errors"
+	"fmt"
+
+	"github.com/ava-labs/avalanchego/utils/wrappers"
+	"github.com/ava-labs/coreth/params"
+	"github.com/ethereum/go-ethereum/common/math"
+)
+
+var ErrInsufficientDynamicFeeWindowLength = errors.New("insufficient length for dynamic fee window")
```
Comment (qdm12) [inline consensus/dummy/dynamic_fee_window.go:16]: nit starting the err name with `ErrDynamicFeeWindow` would be better when searching for errors related to the dynamic fee window, so you could rename this to `ErrDynamicFeeWindowInsufficientLength`

File: consensus/dummy/dynamic_fee_window.go [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+func ParseDynamicFeeWindow(bytes []byte) (DynamicFeeWindow, error) {
+	if len(bytes) < params.DynamicFeeExtraDataSize {
+		return DynamicFeeWindow{}, fmt.Errorf("%w expected at least %d, but found %d",
+			ErrInsufficientDynamicFeeWindowLength,
+			params.DynamicFeeExtraDataSize,
+			len(bytes),
+		)
+	}
+
+	var window DynamicFeeWindow
```
Comment (qdm12) [inline consensus/dummy/dynamic_fee_window.go:34]: supernit you could save a line if using named returns

```suggestion
func ParseDynamicFeeWindow(bytes []byte) (window DynamicFeeWindow, err error) {
	if len(bytes) < params.DynamicFeeExtraDataSize {
		return DynamicFeeWindow{}, fmt.Errorf("%w expected at least %d, but found %d",
			ErrInsufficientDynamicFeeWindowLength,
			params.DynamicFeeExtraDataSize,
			len(bytes),
		)
	}

```

File: consensus/dummy/dynamic_fee_window.go [hunk 11]
```diff
@@ -0,0 +1,92 @@
...
+// DynamicFeeWindow is a window of the last [params.RollupWindow] seconds of gas
+// usage.
+//
+// Index 0 is the oldest entry, and [params.RollupWindow]-1 is the current
+// entry.
+type DynamicFeeWindow [params.RollupWindow]uint64
+
+func ParseDynamicFeeWindow(bytes []byte) (DynamicFeeWindow, error) {
+	if len(bytes) < params.DynamicFeeExtraDataSize {
+		return DynamicFeeWindow{}, fmt.Errorf("%w expected at least %d, but found %d",
```
Comment (qdm12) [inline consensus/dummy/dynamic_fee_window.go:27]: nit just for clarity

```suggestion
		return DynamicFeeWindow{}, fmt.Errorf("%w: expected at least %d bytes but got %d bytes",
```

File: consensus/dummy/dynamic_fee_window_test.go [hunk 12]
```diff
@@ -0,0 +1,246 @@
...
+	"strconv"
+	"testing"
+
+	"github.com/ava-labs/coreth/params"
+	"github.com/ethereum/go-ethereum/common/math"
+	"github.com/stretchr/testify/require"
+)
+
+func TestDynamicFeeWindow_Add(t *testing.T) {
+	tests := []struct {
```
Comment (qdm12) [inline consensus/dummy/dynamic_fee_window_test.go:16]: supernit you could use `map[string]struct{...}` to have the subtest name as the map key - it formats nicely imo the test cases. The only drawback (or not) is that subtests run without order.

File: consensus/dummy/dynamic_fee_window_test.go [hunk 13]
```diff
@@ -0,0 +1,246 @@
+// (c) 2019-2025, Ava Labs, Inc. All rights reserved.
```
Comment (qdm12) [inline consensus/dummy/dynamic_fee_window_test.go:1]:
```suggestion
// (c) 2025, Ava Labs, Inc. All rights reserved.
```

File: consensus/dummy/dynamic_fees.go [hunk 14]
```diff
@@ -83,65 +69,55 @@ func CalcBaseFee(config *params.ChainConfig, parent *types.Header, timestamp uin
...
-			}
-		}
+	// Compute the new state of the gas rolling window.
+	dynamicFeeWindow.Add(parent.GasUsed, parentExtraStateGasUsed, blockGasCost)
 
-		slot := params.RollupWindow - 1 - roll
-		start := slot * wrappers.LongLen
-		updateLongWindow(newRollupWindow, start, addedGas)
+	if timestamp < parent.Time {
+		return nil, nil, fmt.Errorf("cannot calculate base fee for timestamp (%d) prior to parent timestamp (%d)", timestamp, parent.Time)
```
Comment (qdm12) [inline consensus/dummy/dynamic_fees.go:108]: nit why do we use parentheses for the timestamp numbers?

File: consensus/dummy/dynamic_fees.go [hunk 15]
```diff
@@ -83,65 +69,55 @@ func CalcBaseFee(config *params.ChainConfig, parent *types.Header, timestamp uin
...
-				ApricotPhase4MaxBlockGasCost,
-				ApricotPhase4BlockGasCostStep,
-				parent.BlockGasCost,
-				parent.Time, timestamp,
-			).Uint64()
+	// Add in parent's consumed gas
+	var blockGasCost, parentExtraStateGasUsed uint64
+	switch {
+	case isApricotPhase5:
+		// [blockGasCost] has been removed in AP5, so it is left as 0.
```
Comment (qdm12) [inline consensus/dummy/dynamic_fees.go:76]: brackets won't work for local variables

```suggestion
		// `blockGasCost` has been removed in AP5, so it is left as 0.
```

File: consensus/dummy/dynamic_fees.go [hunk 16]
```diff
@@ -83,65 +69,55 @@ func CalcBaseFee(config *params.ChainConfig, parent *types.Header, timestamp uin
...
+			parentExtraStateGasUsed = parent.ExtDataGasUsed.Uint64()
 		}
-
-		// Compute the new state of the gas rolling window.
-		addedGas, overflow := math.SafeAdd(parent.GasUsed, parentExtraStateGasUsed)
-		if overflow {
-			addedGas = math.MaxUint64
+	case isApricotPhase4:
+		// The [blockGasCost] is paid by the effective tips in the block using
+		// the block's value of [baseFee].
```
Comment (qdm12) [inline consensus/dummy/dynamic_fees.go:85]: brackets won't work for local variables

```suggestion
		// The `blockGasCost` is paid by the effective tips in the block using
		// the block's value of `baseFee`.
```

File: consensus/dummy/dynamic_fees.go [hunk 17]
```diff
@@ -83,65 +69,55 @@ func CalcBaseFee(config *params.ChainConfig, parent *types.Header, timestamp uin
...
 
-			// On the boundary of AP3 and AP4 or at the start of a new network, the parent
-			// may not have a populated [ExtDataGasUsed].
-			if parent.ExtDataGasUsed != nil {
-				parentExtraStateGasUsed = parent.ExtDataGasUsed.Uint64()
-			}
-		default:
-			blockGasCost = ApricotPhase3BlockGasFee
+		// At the start of a new network, the parent
+		// may not have a populated [ExtDataGasUsed].
```
Comment (qdm12) [inline consensus/dummy/dynamic_fees.go:79]: nit Those brackets won't work since it's not a global symbol, prefer using backticks

```suggestion
		// may not have a populated `ExtDataGasUsed`.
```

File: consensus/dummy/dynamic_fees.go [hunk 18]
```diff
@@ -83,65 +69,55 @@ func CalcBaseFee(config *params.ChainConfig, parent *types.Header, timestamp uin
...
+			ApricotPhase4TargetBlockRate,
+			ApricotPhase4MinBlockGasCost,
+			ApricotPhase4MaxBlockGasCost,
+			ApricotPhase4BlockGasCostStep,
+			parent.BlockGasCost,
+			parent.Time, timestamp,
+		).Uint64()
+
+		// On the boundary of AP3 and AP4 or at the start of a new network, the parent
+		// may not have a populated [ExtDataGasUsed].
```
Comment (qdm12) [inline consensus/dummy/dynamic_fees.go:96]: nit Those brackets won't work since it's not a global symbol, prefer using backticks

```suggestion
		// may not have a populated `ExtDataGasUsed`.
```

**Review by qdm12 — DISMISSED** [review]

File: consensus/dummy/dynamic_fees.go (hunk 14 above)
Comment (StephenButtolph) [reply consensus/dummy/dynamic_fees.go]: 🤷 that's just how it was. I removed the parens.

File: consensus/dummy/dynamic_fee_window_test.go (hunk 12 above)
Comment (StephenButtolph) [reply consensus/dummy/dynamic_fee_window_test.go:16]: I generally find it easier to debug / work with a test where the order is consistent... So I always use the struct version rather than the map approach... But If people feel strongly I can switch to this.

File: consensus/dummy/dynamic_fee_window.go (hunk 10 above)
Comment (StephenButtolph) [reply consensus/dummy/dynamic_fee_window.go]: I typically avoid using named returns other than for exceptional cases... It doesn't seem like named returns are very common in coreth, so I think it makes sense to keep the line rather than using named returns.

File: consensus/dummy/dynamic_fee_window_test.go (hunk 12 above)
Comment (darioush) [reply consensus/dummy/dynamic_fee_window_test.go:16]: I also prefer the consistent order of arrays here

**Review by darioush — DISMISSED** [review]

File: consensus/dummy/dynamic_fee_window.go [hunk 19]
```diff
@@ -0,0 +1,88 @@
...
+		return DynamicFeeWindow{}, fmt.Errorf("%w: expected at least %d bytes but got %d bytes",
+			ErrDynamicFeeWindowInsufficientLength,
+			params.DynamicFeeExtraDataSize,
+			len(bytes),
+		)
+	}
+
+	var window DynamicFeeWindow
+	for i := range window {
+		offset := i * wrappers.LongLen
```
Comment (ARR4N) [inline consensus/dummy/dynamic_fee_window.go:36]: (subjective) I think `wrappers.LongLen` adds very little while requiring the reader to check its value. We're extracting a `uint64` from bytes, so a raw `8` is an acceptable magic number IMO.

File: consensus/dummy/dynamic_fee_window.go [hunk 20]
```diff
@@ -0,0 +1,88 @@
...
+
+// Add adds the amounts to the most recent entry in the window.
+//
+// If the most recent entry overflows, it is set to [math.MaxUint64].
+func (w *DynamicFeeWindow) Add(amounts ...uint64) {
+	const lastIndex uint = params.RollupWindow - 1
+	w[lastIndex] = add(w[lastIndex], amounts...)
+}
+
+// Shift removes the oldest amount entries from the window and adds amount new
```
Comment (ARR4N) [inline consensus/dummy/dynamic_fee_window.go:50]:
```suggestion
// Shift removes the oldest `amount` entries from the window and adds `amount` new
```

Or rename it to `n` (it's a short function so that's fine, and `n` is self-explanatory). Being an English word it was very confusing to read "removes the oldest amount" (valid sentence thus far) only to then realise that it's a variable.

https://en.wikipedia.org/wiki/Garden-path_sentence

File: consensus/dummy/dynamic_fee_window.go [hunk 21]
```diff
@@ -0,0 +1,88 @@
...
+//
+// If the sum overflows, [math.MaxUint64] is returned.
+func (w *DynamicFeeWindow) Sum() uint64 {
+	return add(0, w[:]...)
+}
+
+func (w *DynamicFeeWindow) Bytes() []byte {
+	bytes := make([]byte, params.DynamicFeeExtraDataSize)
+	for i, v := range w {
+		offset := i * wrappers.LongLen
```
Comment (ARR4N) [inline consensus/dummy/dynamic_fee_window.go:73]: As above re `wrappers.LongLen`.

File: consensus/dummy/dynamic_fee_window_test.go [hunk 22]
```diff
@@ -0,0 +1,246 @@
...
+			},
+		},
+		{
+			amount: 9,
+			expected: DynamicFeeWindow{
+				10,
+			},
+		},
+		{
+			amount:   params.RollupWindow,
```
Comment (ARR4N) [inline consensus/dummy/dynamic_fee_window_test.go:128]: I'd use a raw 10 here (or as another test). What's important is that this is shifting by one more than the last test (i.e. the edge case), not that it's by the actual value of `params.RollupWindow`, even though they're the same thing.

File: consensus/dummy/dynamic_fee_window_test.go [hunk 23]
```diff
@@ -0,0 +1,246 @@
...
+		},
+		{
+			name:   "zero_window",
+			bytes:  make([]byte, params.DynamicFeeExtraDataSize),
+			window: DynamicFeeWindow{},
+		},
+		{
+			name: "truncate_bytes",
+			bytes: []byte{
+				params.DynamicFeeExtraDataSize: 1,
```
Comment (ARR4N) [inline consensus/dummy/dynamic_fee_window_test.go:198]: What is this sorcery?!!!

TIL about this syntax for slice literals 🤯

**Review by ARR4N — APPROVED** [review]
Readability LGTM

File: consensus/dummy/dynamic_fee_window.go (hunk 19 above)
Comment (StephenButtolph) [reply consensus/dummy/dynamic_fee_window.go:36]: Hmmm I feel we should keep using `wrappers.LongLen` because that aligns with the definition of:
```golang
const DynamicFeeExtraDataSize = RollupWindow * wrappers.LongLen
```
